flink-batch-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCInputFormat.java

```diff
@@ -139,12 +184,13 @@ public void close() throws IOException {
 	@Override
 	public boolean reachedEnd() throws IOException {
 		try {
-			if (resultSet.isLast()) {
-				close();
+			if (!hasNext || resultSet.isLast()) {
```

Comment:
I think we should not use `resultSet.isLast()`. 
Instead we can call `ResultSet.next()` initially in `open()` and at the end of `nextRecord()` and use the `hasNext` flag in `reachedEnd()`.